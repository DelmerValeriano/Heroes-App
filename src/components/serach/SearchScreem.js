import { useLocation, useNavigate } from "react-router-dom";
import React, { useMemo } from 'react'


import { useForm } from "../../hook/useForm";
import { getHeroesByName } from "../../selectors/getHeroesByName";
import { HeroCard } from "../hero/HeroCard";
import queryString from 'query-string'




export const SearchScreem = () => {



  const navigate =useNavigate();
  const location =useLocation();
 
  const {q=''} = queryString.parse(location.search);


  const [formValue,handleInputChange] = useForm({
    BuscarText:q

  });

  const {BuscarText} =formValue;

  const heroesFileted =useMemo(()=>  getHeroesByName(q),[q]);

  const handleSearch = (e) => {
    e.preventDefault();
   
    navigate(`?q=${BuscarText}`);

  };


  return (
    <>
      <h1>Busqueda</h1>
      <hr/>
     

      <div className="row">
       

          <div className="col-5">
              <h4>Buscar</h4>
              <hr/>
              

              <form onSubmit={handleSearch}>
                <input 
                  type="text" 
                  className="form-control"
                  placeholder="Burcar un Heroe"
                  name="BuscarText"
                  autoComplete="off"
                  value={BuscarText}
                  onChange={handleInputChange}
                />
                <div className="d-grid gap-2">
                  <button
                    type="submit"
                    className="btn btn-outline-primary mt-2"
                  >
                      Buscar...
                  </button>

                </div>
                
              </form>
          </div>

          <div className="col-7">
            <h4>Resultados</h4>
            <hr/>
            {
                (q==='')
                ? <div className="alert alert-info">Buscar un heroe</div>
                 : (heroesFileted.length ===0) 
                  && <div className="alert alert-danger">No hay resultados: {q}</div>

            }


            {
              heroesFileted.map(hero=>(
                <HeroCard
                  key={hero.id}
                  {...hero}
                
                />
              ))
            }



          </div>
       </div>    
  
    </>
  )
}
