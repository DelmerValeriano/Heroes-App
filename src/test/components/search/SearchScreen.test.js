import { mount } from "enzyme"
import { SearchScreem } from "../../../components/serach/SearchScreem"
import React from 'react';
import { MemoryRouter } from "react-router-dom";


const mockNavigate =jest.fn();
jest.mock('react-router-dom',()=>({
    ...jest.requireActual('react-router-dom'),
    useNavigate:()=>mockNavigate,

}));


describe('pruebas en SearchScreem', () => {  

    test('debe de mostrarse correctamente con valores por defecto', () => {  
        const wrapper = mount(
            <MemoryRouter initialEntries={['/']}>
                 <SearchScreem/>
            </MemoryRouter>
        );
        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('.alert-info').text().trim()).toBe('Buscar un heroe');
    });

    test('debe de mostar a batman y el input con el valor del queryString', () => {  

        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=batman']}>
                 <SearchScreem/>
            </MemoryRouter>
        );

        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('input').prop('value')).toBe('batman');
    });

    test('debe de mostrar un error si no hay resultados', () => {  
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=batman123']}>
                 <SearchScreem/>
            </MemoryRouter>
        );

        expect(wrapper.find('.alert-danger').text().trim()).toBe('No hay resultados: batman123');
    });

    test('debe de llamar al navigate a la nueva pantalla', () => { 
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search']}>
                 <SearchScreem/>
            </MemoryRouter>
        );

        wrapper.find('input').simulate('change',{
            target: {
                name:'BuscarText',
                value: 'batman',
            }
        });
        wrapper.find('form').prop('onSubmit')({
            preventDefault: ()=>{}

        })
        expect(mockNavigate).toHaveBeenCalledWith('?q=batman');




    });


})
            
