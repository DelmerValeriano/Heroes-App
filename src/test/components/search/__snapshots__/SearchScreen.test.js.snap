// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pruebas en SearchScreem debe de mostar a batman y el input con el valor del queryString 1`] = `
Array [
  <h1>
    Busqueda
  </h1>,
  <hr />,
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Buscar
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="BuscarText"
          onChange={[Function]}
          placeholder="Burcar un Heroe"
          type="text"
          value="batman"
        />
        <div
          className="d-grid gap-2"
        >
          <button
            className="btn btn-outline-primary mt-2"
            type="submit"
          >
            Buscar...
          </button>
        </div>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Resultados
      </h4>
      <hr />
      <div
        className="col"
      >
        <div
          className="card"
        >
          <div
            className="row no-gutters"
          >
            <div
              className="col-4"
            >
              <img
                alt="Batman"
                className="card-img"
                src="/assets/dc-batman.jpg"
              />
            </div>
            <div
              className="col-8"
            >
              <div
                className="card-body"
              >
                <h5
                  className="card-title"
                >
                  Batman
                </h5>
                <p
                  className="card-text"
                >
                  Bruce Wayne
                </p>
                <p
                  className="text-muted"
                >
                  <small
                    className="text-muted"
                  >
                      
                    Detective Comics #27
                      
                  </small>
                </p>
                <a
                  href="/hero/dc-batman"
                  onClick={[Function]}
                >
                  Mass...
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
]
`;

exports[`pruebas en SearchScreem debe de mostrarse correctamente con valores por defecto 1`] = `
Array [
  <h1>
    Busqueda
  </h1>,
  <hr />,
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Buscar
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="BuscarText"
          onChange={[Function]}
          placeholder="Burcar un Heroe"
          type="text"
          value=""
        />
        <div
          className="d-grid gap-2"
        >
          <button
            className="btn btn-outline-primary mt-2"
            type="submit"
          >
            Buscar...
          </button>
        </div>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Resultados
      </h4>
      <hr />
      <div
        className="alert alert-info"
      >
        Buscar un heroe
      </div>
    </div>
  </div>,
]
`;
